File: web/src/esp32/protocol.ts
/**
 * PC <-> ESP32 serial protocol — must stay byte-for-byte in sync with
 * src/bcihand/communication/protocol.py and
 * firmware/esp32/include/protocol.h. See those files' docstrings for the
 * full frame reference:
 *
 * Frames browser -> ESP32:  CMD:OPEN\n  CMD:CLOSE\n  CMD:HOLD\n  HEARTBEAT\n
 * Frames ESP32 -> browser:  ACK:<CMD>\n  STATE:<STATE>\n  HB_ACK\n
 */
import type { Command } from "../core/types";

export const CMD_PREFIX = "CMD:";
export const ACK_PREFIX = "ACK:";
export const STATE_PREFIX = "STATE:";
export const HEARTBEAT = "HEARTBEAT";
export const HEARTBEAT_ACK = "HB_ACK";

const encoder = new TextEncoder();

export function encodeCommand(command: Command): Uint8Array<ArrayBuffer> {
  return encoder.encode(`${CMD_PREFIX}${command}\n`);
}

export function encodeHeartbeat(): Uint8Array<ArrayBuffer> {
  return encoder.encode(`${HEARTBEAT}\n`);
}

export type DecodedFrame =
  | { kind: "ACK"; payload: string }
  | { kind: "STATE"; payload: string }
  | { kind: "HB_ACK"; payload: "" };

export function decodeLine(rawLine: string): DecodedFrame | null {
  const line = rawLine.trim();
  if (!line) return null;
  if (line === HEARTBEAT_ACK) return { kind: "HB_ACK", payload: "" };
  if (line.startsWith(ACK_PREFIX)) return { kind: "ACK", payload: line.slice(ACK_PREFIX.length) };
  if (line.startsWith(STATE_PREFIX)) return { kind: "STATE", payload: line.slice(STATE_PREFIX.length) };
  return null;
}

File: src/bcihand/communication/protocol.py
"""PC <-> ESP32 serial protocol (Mode A, USB serial).

Plain-text, newline-terminated, human-readable-over-a-terminal frames — easy
to debug with any serial monitor while this is still being bench-tested.

Frames PC -> ESP32:
    CMD:OPEN\n
    CMD:CLOSE\n
    CMD:HOLD\n
    HEARTBEAT\n

Frames ESP32 -> PC:
    ACK:<CMD>\n      (e.g. "ACK:OPEN")
    STATE:<STATE>\n  (current firmware state, e.g. "STATE:HOLD")
    HB_ACK\n

This exact framing is mirrored in firmware/esp32/src/comms.cpp /
command_parser.cpp — keep the two in sync if you change it. See
docs/ESP32_SETUP.md.
"""
from __future__ import annotations

CMD_PREFIX = "CMD:"
ACK_PREFIX = "ACK:"
STATE_PREFIX = "STATE:"
HEARTBEAT = "HEARTBEAT"
HEARTBEAT_ACK = "HB_ACK"

VALID_COMMANDS = ("OPEN", "CLOSE", "HOLD")


def encode_command(command_name: str) -> bytes:
    if command_name not in VALID_COMMANDS:
        raise ValueError(f"Invalid command: {command_name}")
    return f"{CMD_PREFIX}{command_name}\n".encode("ascii")


def encode_heartbeat() -> bytes:
    return f"{HEARTBEAT}\n".encode("ascii")


def decode_line(line: str) -> tuple[str, str] | None:
    """Returns (kind, payload) for a received line, or None if unrecognized.
    kind is one of: "ACK", "STATE", "HB_ACK"."""
    line = line.strip()
    if not line:
        return None
    if line == HEARTBEAT_ACK:
        return ("HB_ACK", "")
    if line.startswith(ACK_PREFIX):
        return ("ACK", line[len(ACK_PREFIX):])
    if line.startswith(STATE_PREFIX):
        return ("STATE", line[len(STATE_PREFIX):])
    return None
